### PR TITLE
Fix for snapshot stuck deletion

### PR DESCRIPTION
**Pull request by hemalathagajendran**
When multiple snapshots are taken within snap window for version2 volume then snapshot deletion was getting stuck. This fix resolves this snapshot deletion stuck issue introduced with shallow copy feature.

Please check the type of change your PR introduces:
- [ x] Bugfix
- [ ] Feature Enhancement
- [ ] Test Automation 
- [ ] Code Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Community Operator listing 
- [ ] Other (please describe): 

## What is the current behavior?

- Snapshot deletion is getting stuck when multiple snapshots are taken within snap window for version2 volume

## What is the new behavior?

- Snapshot deletion works fine if multiple snapshots are taken within snap window for version2 volume

## How risky is this change?
- [ ] Small, isolated change
- [ x] Medium, requires regression testing
- [ ] Large, requires functional and regression testing

### CODE REVIEW

File: driver/csiplugin/controllerserver.go [hunk 1]
```diff
@@ -2098,7 +2102,7 @@ func (cs *ScaleControllerServer) DeleteVolume(ctx context.Context, req *csi.Dele
 	return &csi.DeleteVolumeResponse{}, nil
 }
 
-func (cs *ScaleControllerServer) DeleteShallowCopyRefPath (ctx context.Context, FilesystemName, FilesetName, ShallowCopyRefPath, storageClassType string, conn connectors.SpectrumScaleConnector) (bool, error){
+func (cs *ScaleControllerServer) DeleteShallowCopyRefPath (ctx context.Context, FilesystemName, FilesetName, ShallowCopyRefPath, storageClassType, independentFileset, snapshotName string, conn connectors.SpectrumScaleConnector) (bool, error){
```
Comment (badri-pathak) [inline driver/csiplugin/controllerserver.go:2105]: Since Return bool is not being used. Can we remove this?

**Comment by hemalathagajendran** [comment]
quay.io/hemalatha_gajendran/snapshot_stuck_issue_final - Use this latest one

File: driver/csiplugin/controllerserver.go (hunk 1 above)
Comment (hemalathagajendran) [reply driver/csiplugin/controllerserver.go:2105]: Done

**Comment by hemalathagajendran** [comment]
Image: quay.io/hemalatha_gajendran/remote_fsfix:latest (Fix for remotefs issue)

**Review by saurabhwani5 — APPROVED** [review]

**Review by saurabhwani5 — CHANGES_REQUESTED** [review]
clone of shallow copy is not working on ocp

**Comment by hemalathagajendran** [comment]
> clone of shallow copy is not working on ocp

quay.io/hemalatha_gajendran/shallowcopy_multiple_fixes_final:latest

**Review by saurabhwani5 — DISMISSED** [review]

File: driver/csiplugin/controllerserver.go [hunk 2]
```diff
@@ -1600,69 +1605,65 @@ func (cs *ScaleControllerServer) validateSnapId(ctx context.Context, scaleVol *s
...
-						}
-					}
-
-					if !isSamefsetType {
-						klog.Errorf("[%s] Filesettype is not same for both source snapshot and new volume", loggerId)
-						return false
-					}
-				}
+          				isSamefsetType := false
+          				if newvolume.FilesetType == independentFileset {
```
Comment (badri-pathak) [inline driver/csiplugin/controllerserver.go:1648]: Can we combine this like?
( newvolume.FilesetType == dependentFileset && sourcesnapshot.VolType == FILE_DEPENDENTFILESET_VOLUME ) || ( newvolume.FilesetType == independentFileset && sourcesnapshot.VolType == FILE_INDEPENDENTFILESET_VOLUME)

Comment (badri-pathak) [inline driver/csiplugin/controllerserver.go:1648]: Can be updated for multiple if conditions.

File: driver/csiplugin/controllerserver.go [hunk 3]
```diff
@@ -2079,50 +2084,59 @@ func (cs *ScaleControllerServer) DeleteVolume(ctx context.Context, req *csi.Dele
...
 	err := conn.DeleteDirectory(ctx, FilesystemName, shallowCopyRefCompletePath, false)
 	if err != nil {
 		if strings.Contains(err.Error(), "EFSSG0264C") ||
 			strings.Contains(err.Error(), "does not exist") { // directory is already deleted
 			isShallowCopyRefPathDeleted = true
-		} else {
-			return false, status.Error(codes.Internal, fmt.Sprintf("unable to Delete shallow copy reference Dir using FS [%v] Error [%v]", FilesystemName, err))
-		}
+		}else{
+			return status.Error(codes.Internal, fmt.Sprintf("unable to Delete shallow copy reference Dir using FS [%v] Error [%v]", FilesystemName, err))
```
Comment (badri-pathak) [inline driver/csiplugin/controllerserver.go:2102]: %v is not recommended for string. Sometimes it gets overwrite with other type. Can use %s in all other places like clusterID

File: driver/csiplugin/controllerserver.go [hunk 4]
```diff
@@ -1600,69 +1605,65 @@ func (cs *ScaleControllerServer) validateSnapId(ctx context.Context, scaleVol *s
...
+            					if sourcesnapshot.VolType == FILE_INDEPENDENTFILESET_VOLUME{
+              						isSamefsetType = true
+            					}
+          				}else if newvolume.FilesetType == dependentFileset{
+            					if sourcesnapshot.VolType == FILE_DEPENDENTFILESET_VOLUME{
+              						isSamefsetType = true
+             					}
+          				}
+
+          				if !isSamefsetType {
```
Comment (badri-pathak) [inline driver/csiplugin/controllerserver.go:1658]: This can also be clubbed with above if.

File: driver/csiplugin/controllerserver.go [hunk 5]
```diff
@@ -1974,16 +1975,20 @@ func (cs *ScaleControllerServer) DeleteVolume(ctx context.Context, req *csi.Dele
...
-					shallowCopyRefPath = fmt.Sprintf("%s/%s", volPath[0], volPath[2])
-				}
-			}
-		}
+	var snapshotName string
+  	var independentFileset string
+	if volumeIdMembers.VolType == FILE_SHALLOWCOPY_VOLUME{
+    		if relPath != "" && strings.Contains(relPath, ".snapshots"){
+      			volPath := strings.Split(relPath, "/")
+        		if len(volPath) > 2{
```
Comment (badri-pathak) [inline driver/csiplugin/controllerserver.go:1983]: can we add else when length is less than or equal to 2 with error or just log, or what will happen when if doesn't meet

File: driver/csiplugin/controllerserver.go (hunk 2 above)
Comment (hemalathagajendran) [reply driver/csiplugin/controllerserver.go]: Done

File: driver/csiplugin/controllerserver.go (hunk 3 above)
Comment (hemalathagajendran) [reply driver/csiplugin/controllerserver.go]: Done

File: driver/csiplugin/controllerserver.go (hunk 4 above)
Comment (hemalathagajendran) [reply driver/csiplugin/controllerserver.go]: Done

File: driver/csiplugin/controllerserver.go (hunk 5 above)
Comment (hemalathagajendran) [reply driver/csiplugin/controllerserver.go:1973]: Done

**Review by badri-pathak — APPROVED** [review]

**Review by saurabhwani5 — APPROVED** [review]

**Pull request merged**